File: numba/float_matrix_multiplication.py
import time
import numpy as np
from numba import cuda

@cuda.jit
def matrix_mul(a, b, c, n):
    tid = cuda.grid(1)
    i = tid // n
    j = tid % n
    if i < n and j < n:
        sum = 0
        for k in range(n):
            sum += a[i * n + k] * b[k * n + j]
        c[i * n + j] = sum

def bench9():
    # 1 is for precompilation
    sizes = [1, 16, 64, 256, 512, 1024, 2048, 4096]
    results = []

    for n in sizes:
        total = n * n

        a = np.array([(i * 12345.6789) % 1000 for i in range(total)], dtype=float)
        b = np.array([(i * 98765.4321) % 1000 for i in range(total)], dtype=float)
        c = np.zeros(total, dtype=float)

        d_a = cuda.to_device(a)
        d_b = cuda.to_device(b)
        d_c = cuda.device_array_like(c)

        threads_per_block = 256
        blocks = (total + threads_per_block - 1) // threads_per_block

        start = time.perf_counter()
        matrix_mul[blocks, threads_per_block](d_a, d_b, d_c, n)
        d_c.copy_to_host(c)
        elapsed = (time.perf_counter() - start) * 1000


        if n != 1:
            print(f"bench9 - Size: {n}x{n} | Time: {elapsed:.3f} ms")
            results.append({
                'test': f'bench9_{n}x{n}',
                'size': n,
                'gpu_time_ms': elapsed
            })

    return results

def bench10():
    # 1 is for precompilation
    sizes = [1, 16, 64, 256, 512, 1024, 2048, 4096]
    results = []

    for n in sizes:
        total = n * n

        a = np.array([(i * 12345.6789) % 1000 for i in range(total)], dtype=np.float32)
        b = np.array([(i * 98765.4321) % 1000 for i in range(total)], dtype=np.float32)
        c = np.zeros(total, dtype=np.float32)

        d_a = cuda.to_device(a)
        d_b = cuda.to_device(b)
        d_c = cuda.device_array_like(c)

        threads_per_block = 256
        blocks = (total + threads_per_block - 1) // threads_per_block

        start = time.perf_counter()
        matrix_mul[blocks, threads_per_block](d_a, d_b, d_c, n)
        d_c.copy_to_host(c)
        elapsed = (time.perf_counter() - start) * 1000


        if n != 1:
            print(f"bench10 - Size: {n}x{n} | Time: {elapsed:.3f} ms")
            results.append({
                'test': f'bench10_{n}x{n}',
                'size': n,
                'gpu_time_ms': elapsed
            })

    return results

File: rust/src/float_matrix_multiplication.rs
use cuda::dmem::{Buffer, DSend, DeepCopy};
use cuda::{gpu};
use std::time::Instant;
use crate::BenchResults;

#[kernel]
unsafe fn matrix_mul(mut result: Buffer<f64>,
                         a: &[f64], b: &[f64], width: usize, height: usize) {
    let tid = gpu::global_tid_x() as usize;
    let i = tid / height;
    let j = tid % height;
    let mut sum:f64 = 0.0;
    for k in 0..width {
        sum = sum + a[i * width + k] * b[k * height + j];
    }
    result.set(i * height + j, sum);
}

#[kernel]
unsafe fn matrix_mul_f32(mut result: Buffer<f32>,
                     a: &[f32], b: &[f32], width: usize, height: usize) {
    let tid = gpu::global_tid_x() as usize;
    let i = tid / height;
    let j = tid % height;
    let mut sum:f32 = 0.0;
    for k in 0..width {
        sum = sum + a[i * width + k] * b[k * height + j];
    }
    result.set(i * height + j, sum);
}

pub fn bench9() -> Vec<BenchResults> {
    match matrix_mul.pre_compile() {
        Ok(_) => {},
        Err(e) => {
            println!("Kernel compilation failed: {:?}", e);
            panic!("Kernel compilation failed");
        }
    }

    let mut results = Vec::new();
    let sizes: [usize;7] = [64, 128, 256, 512, 1024, 2048, 4096];

    for n in sizes {
        let total = n * n;

        // Generate deterministic pseudo-random matrix data
        let mut a = vec![0.0; n * n];
        let mut b = vec![0.0; n * n];
        for i in 0..n * n {
            a[i] = i as f64 * 12345.6789 % 1000.0;
            b[i] = i as f64 * 98765.4321 % 1000.0;
        }

        let mut d_a = a.as_slice().to_device().unwrap();
        let mut d_b = b.as_slice().to_device().unwrap();
        let buf_c = Buffer::<f64>::alloc(total).unwrap();
        let mut d_buf_c = buf_c.to_device().unwrap();

        let threads_per_block = 256;
        let blocks = (n * n + threads_per_block - 1) / threads_per_block;

        let start = Instant::now();
        let mut d_width = n.to_device().unwrap();
        let mut d_height = n.to_device().unwrap();
        match matrix_mul.launch_with_dptr(
            threads_per_block,
            blocks,
            &mut d_buf_c,
            &mut d_a,
            &mut d_b,
            &mut d_width,
            &mut d_height
        ) {
            Ok(_) => {}
            Err(e) => {
                eprintln!("Error launching matrix mul kernel: {:?}", e);
            }
        }

        let x = buf_c.retrieve().unwrap();
        let elapsed_gpu_ms = start.elapsed().as_secs_f64() * 1000.0;

        results.push(BenchResults {
            gpu_time: elapsed_gpu_ms,
            test: format!("{}x{}", n, n),
            size: n,
        });
    }

    println!("Bench 9 Results:");
    for result in &results {
        println!("Test: {} | GPU Time: {:.3} ms", result.test, result.gpu_time);
    }

    results
}

pub fn bench10() -> Vec<BenchResults> {
    match matrix_mul_f32.pre_compile() {
        Ok(_) => {},
        Err(e) => {
            println!("Kernel compilation failed: {:?}", e);
            panic!("Kernel compilation failed");
        }
    }

    let mut results = Vec::new();
    let sizes: [usize;7] = [64, 128, 256, 512, 1024, 2048, 4096];

    for n in sizes {
        let total = n * n;

        // Generate deterministic pseudo-random matrix data
        let mut a = vec![0.0; n * n];
        let mut b = vec![0.0; n * n];
        for i in 0..n * n {
            a[i] = i as f32 * 12345.6789 % 1000.0;
            b[i] = i as f32 * 98765.4321 % 1000.0;
        }

        let mut d_a = a.as_slice().to_device().unwrap();
        let mut d_b = b.as_slice().to_device().unwrap();
        let buf_c = Buffer::<f32>::alloc(total).unwrap();
        let mut d_buf_c = buf_c.to_device().unwrap();

        let threads_per_block = 256;
        let blocks = (n * n + threads_per_block - 1) / threads_per_block;

        let start = Instant::now();
        let mut d_width = n.to_device().unwrap();
        let mut d_height = n.to_device().unwrap();
        match matrix_mul_f32.launch_with_dptr(
            threads_per_block,
            blocks,
            &mut d_buf_c,
            &mut d_a,
            &mut d_b,
            &mut d_width,
            &mut d_height
        ) {
            Ok(_) => {}
            Err(e) => {
                eprintln!("Error launching matrix mul kernel: {:?}", e);
            }
        }

        let x = buf_c.retrieve().unwrap();
        let elapsed_gpu_ms = start.elapsed().as_secs_f64() * 1000.0;

        results.push(BenchResults {
            gpu_time: elapsed_gpu_ms,
            test: format!("{}x{}", n, n),
            size: n,
        });
    }

    println!("Bench 10 Results:");
    for result in &results {
        println!("Test: {} | GPU Time: {:.3} ms", result.test, result.gpu_time);
    }

    results
}